internal/registration/registration_test.go

```diff
@@ -0,0 +1,172 @@
+package registration_test
+
+import (
+	"fmt"
+	"io/ioutil"
+	osUtil "os"
+	"time"
+
+	gomock "github.com/golang/mock/gomock"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	pb "github.com/redhatinsights/yggdrasil/protocol"
+
+	"github.com/jakub-dzon/k4e-device-worker/internal/configuration"
+	"github.com/jakub-dzon/k4e-device-worker/internal/datatransfer"
+	"github.com/jakub-dzon/k4e-device-worker/internal/hardware"
+	"github.com/jakub-dzon/k4e-device-worker/internal/heartbeat"
+	"github.com/jakub-dzon/k4e-device-worker/internal/os"
+	"github.com/jakub-dzon/k4e-device-worker/internal/registration"
+	"github.com/jakub-dzon/k4e-device-worker/internal/workload"
+)
+
+const (
+	deviceConfigName = "device-config.json"
+)
+
+var _ = Describe("Registration", func() {
+
+	var (
+		datadir        string
+		mockCtrl       *gomock.Controller
+		wkManager      *workload.WorkloadManager
+		wkwMock        *workload.MockWorkloadWrapper
+		dispatcherMock *registration.MockDispatcherClient
+		configManager  *configuration.Manager
+		hb             *heartbeat.Heartbeat
+		hw             = &hardware.Hardware{}
+		monitor        = &datatransfer.Monitor{}
+		os             = &os.OS{}
+		err            error
+	)
+
+	BeforeEach(func() {
+		datadir, err = ioutil.TempDir("", "registrationTest")
+		Expect(err).ToNot(HaveOccurred())
+
+		mockCtrl = gomock.NewController(GinkgoT())
+
+		wkwMock = workload.NewMockWorkloadWrapper(mockCtrl)
+		wkwMock.EXPECT().Init().Return(nil).AnyTimes()
+
+		dispatcherMock = registration.NewMockDispatcherClient(mockCtrl)
+
+		wkManager, err = workload.NewWorkloadManagerWithParams(datadir, wkwMock)
+		Expect(err).NotTo(HaveOccurred(), "Cannot start the Workload Manager")
+
+		configManager = configuration.NewConfigurationManager(datadir)
+
+		hb = heartbeat.NewHeartbeatService(dispatcherMock,
+			configManager,
+			wkManager,
+			hw,
+			monitor)
+
+	})
+
+	AfterEach(func() {
+		mockCtrl.Finish()
+		_ = osUtil.Remove(datadir)
+	})
+
+	Context("RegisterDevice", func() {
+
+		RegistrationMatcher := func() gomock.Matcher {
+			return regMatcher{}
+		}
+
+		It("Work as expected", func() {
+
+			// given
+			reg := registration.NewRegistration(
+				hw, os, dispatcherMock, configManager, hb, wkManager, monitor)
+
+			// then
+			dispatcherMock.EXPECT().Send(gomock.Any(), RegistrationMatcher()).Times(1)
+
+			//  when
+			reg.RegisterDevice()
+		})
+
+		It("Try to re-register", func() {
+
+			// given
+			reg := registration.NewRegistration(
+				hw, os, dispatcherMock, configManager, hb, wkManager, monitor)
+			reg.RetryAfter = 1
+
+			// then
+			dispatcherMock.EXPECT().Send(gomock.Any(), gomock.Any()).Return(
+				nil, fmt.Errorf("failed")).Times(1)
+			dispatcherMock.EXPECT().Send(gomock.Any(), RegistrationMatcher()).Times(1)
+
+			//  when
+			reg.RegisterDevice()
+			time.Sleep(2 * time.Second)
```

Comment:
Why do we need to wait? It always causes flaky behavior in CI. Maybe we could add some logic on wait_on_register?

Reply:
Ok, make sense what you're suggesting, I'll implement that

Reply:
waiting for an event/action can be done by using https://onsi.github.io/gomega/#making-asynchronous-assertions
not sure if that will fit here, since the condition to wait for isn't public.